File: backend/services/core/monitoring/sentry.py
"""
Sentry configuration for error tracking and monitoring.
"""
import os
import sentry_sdk
from sentry_sdk.integrations.django import DjangoIntegration
from sentry_sdk.integrations.celery import CeleryIntegration
from sentry_sdk.integrations.redis import RedisIntegration


def init_sentry(environment: str = 'development'):
    """
    Initialize Sentry error tracking.
    
    Args:
        environment: The environment to configure for (development, staging, production)
    """
    sentry_dsn = os.getenv('SENTRY_DSN')
    
    if not sentry_dsn:
        # Sentry is optional - only initialize if DSN is provided
        return
    
    sentry_sdk.init(
        dsn=sentry_dsn,
        integrations=[
            DjangoIntegration(),
            CeleryIntegration(),
            RedisIntegration(),
        ],
        traces_sample_rate=0.1 if environment == 'production' else 1.0,
        send_default_pii=False,  # Don't send PII in production
        environment=environment,
        # Performance Monitoring
        profiles_sample_rate=0.1 if environment == 'production' else 1.0,
        # Release tracking
        release=os.getenv('SENTRY_RELEASE', 'unknown'),
    )


def get_sentry_user_context(user):
    """
    Get user context for Sentry.
    
    Args:
        user: Django User object
    
    Returns:
        dict: User context for Sentry
    """
    if not user or not user.is_authenticated:
        return {}
    
    return {
        'id': user.id,
        'username': user.username,
        'email': user.email,
        'ip_address': '{{auto}}',
    }


def set_sentry_user_context(user):
    """
    Set the current user context in Sentry.
    
    Args:
        user: Django User object
    """
    sentry_sdk.set_user(get_sentry_user_context(user))


def capture_exception(exc, level='error', **kwargs):
    """
    Capture an exception in Sentry.
    
    Args:
        exc: Exception to capture
        level: Severity level (fatal, error, warning, info, debug)
        **kwargs: Additional context
    """
    sentry_sdk.capture_exception(exc)


def capture_message(message, level='info', **kwargs):
    """
    Capture a message in Sentry.
    
    Args:
        message: Message to capture
        level: Severity level (fatal, error, warning, info, debug)
        **kwargs: Additional context
    """
    sentry_sdk.capture_message(message, level)


def add_breadcrumb(message, category='info', level='info', **data):
    """
    Add a breadcrumb to Sentry.
    
    Args:
        message: Breadcrumb message
        category: Breadcrumb category
        level: Severity level
        **data: Additional data
    """
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data,
    )

File: frontend/src/config/sentry.ts
/**
 * Sentry configuration for frontend error tracking.
 * This file is imported in main.tsx
 */

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';

export const initSentry = () => {
  const sentryDSN = import.meta.env.VITE_SENTRY_DSN;
  const environment = import.meta.env.MODE;

  if (!sentryDSN) {
    console.warn('Sentry DSN not configured');
    return;
  }

  Sentry.init({
    dsn: sentryDSN,
    integrations: [
      new BrowserTracing({
        // Set sampling rate for performance monitoring
        tracingOrigins: ['localhost', /^\//],
        routingInstrumentation: Sentry.reactRouterV6Instrumentation(
          window.history
        ),
      }),
      new Sentry.Replay({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],
    // Set tracesSampleRate to 1.0 to capture 100% of transactions for performance monitoring.
    // We recommend adjusting this value in production
    tracesSampleRate: environment === 'production' ? 0.1 : 1.0,
    environment,
    release: import.meta.env.VITE_APP_VERSION || 'unknown',

    // Capture Replay for 10% of all sessions,
    // plus, capture 100% of sessions with an error
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,

    // Don't send PII data
    sendDefaultPii: false,

    // Allowed URLs for error tracking
    allowUrls: [
      /https?:\/\/(localhost|127\.0\.0\.1)/,
      /https?:\/\/yourdomain\.com/,
    ],
  });
};

export const setSentryUser = (user: any) => {
  if (!user) {
    Sentry.setUser(null);
    return;
  }

  Sentry.setUser({
    id: user.id,
    username: user.username,
    email: user.email,
    ip_address: '{{auto}}',
  });
};

export const captureSentryException = (
  error: Error,
  level: 'fatal' | 'error' | 'warning' | 'info' = 'error'
) => {
  Sentry.captureException(error, { level });
};

export const captureSentryMessage = (
  message: string,
  level: 'fatal' | 'error' | 'warning' | 'info' = 'info'
) => {
  Sentry.captureMessage(message, level);
};

export const addBreadcrumb = (
  message: string,
  category: string = 'user-action',
  level: 'fatal' | 'error' | 'warning' | 'info' = 'info',
  data?: Record<string, any>
) => {
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  });
};

export const withSentryErrorBoundary = (Component: React.ComponentType<any>) => {
  return Sentry.withErrorBoundary(Component, {
    fallback: (
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          height: '100vh',
          backgroundColor: '#f5f5f5',
        }}
      >
        <h1>Oops! Something went wrong</h1>
        <p>Our team has been notified. Please try refreshing the page.</p>
        <button onClick={() => window.location.reload()}>
          Refresh Page
        </button>
      </div>
    ),
  });
};
